src/glue/jobs/json_to_parquet.py

```diff
@@ -182,11 +183,86 @@ def drop_deleted_healthkit_data(
     )
     return table_with_deleted_samples_removed
 
+def archive_existing_datasets(
+        glue_context: GlueContext,
+        bucket: str,
+        key_prefix: str,
+        workflow_run_id: str,
+        delete_upon_completion: bool
+    ) -> None:
+    """
+    Archives existing datasets in S3 by copying them to a timestamped subfolder
+    within an "archive" folder. The format of the timestamped subfolder is:
+
+        {year}_{month}_{day}_{workflow_run_id}
+
+    If the dataset does not exist at `key_prefix` or the dataset is empty,
+    no action will be taken.
+
+    Args:
+        glue_context (GlueContext): The Glue context
+        bucket (str): The name of the S3 bucket containing the dataset to archive.
+        key_prefix (str): The S3 key prefix of the objects to be archived.
+            Ensure that this ends with a trailing "/" to avoid matching
+            child datasets.
+        workflow_run_id (str): The Glue workflow run ID.
+        delete_upon_completion (bool): Whether to delete the dataset after archiving.
+
+    Returns:
+        None
+
+    Raises:
+        N/A
+    """
+    logger = glue_context.get_logger()
+    s3_client = boto3.client("s3")
+    response = s3_client.list_objects_v2(
+            Bucket=bucket,
+            Prefix=key_prefix
+    )
+    if "Contents" not in response:
+        return
+    formatted_datetime = ( # e.g., 2023_06_28
+            str(datetime.now())
+            .split(" ")[0]
+            .replace('-', '_')
+    )
+    this_archive_collection = "_".join([formatted_datetime, workflow_run_id])
+    source_objects = [
+            obj for obj in response["Contents"] if obj["Size"] > 0]
+    objects_to_delete = []
+    for source_object in source_objects:
```

Comment:
Nit: This is trivial, but you are looping through this list twice.  

I think its fine - what is your rationale for doing this?

Reply:
It's prettier than
```
for source_object in [obj for obj in response["Contents"] if obj["Size"] > 0]
```